### PR TITLE
Restrict view #46

### PR DESCRIPTION
**Pull request by mknblch**
The commit adds restriction to label-, relation-, and property-types. The transaction handling has also been changed according to the PR comments. Now every method which uses the neo4j core api is wrapped into its own transaction.

### CODE REVIEW

**Comment by jexp** [comment]
A general comment, without looking at the code, it will probably slow down quite a lot to wrap each method. 

Going forward the transaction control should be outside the graph view and the graph view is just participating in the running tx that's started at thread start and finished at thread end.

File: core/src/main/java/org/neo4j/graphalgo/core/neo4jview/RelationIteratorImpl.java [hunk 1]
```diff
@@ -1,27 +1,71 @@
...
+    private final RelationCursor cursor;
+    private final long originalNodeId;
+
+    RelationIteratorImpl(Graph graph, GraphDatabaseAPI api, int sourceNodeId, Direction direction, int relationTypeId) throws EntityNotFoundException {
+        this.graph = graph;
+        this.api = api;
+        transaction = api.beginTx();
+        statement = api.getDependencyResolver()
+                .resolveDependency(ThreadToStatementContextBridge.class)
+                .get();
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/neo4jview/RelationIteratorImpl.java]: this works if we assume that the iterators are only called within one thread, at minimum we have to document that in the GraphView

File: core/src/main/java/org/neo4j/graphalgo/core/neo4jview/GraphView.java [hunk 2]
```diff
@@ -10,98 +10,89 @@
...
             }
-        } catch (EntityNotFoundException e) {
-            throw new RuntimeException(e);
-        }
+            return null;
+        });
     }
 
     @Override
     public Iterator<WeightedRelationCursor> weightedRelationIterator(int nodeId, Direction direction) {
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/neo4jview/GraphView.java]: document on all iterator methods that their resulting iterators can only be used safely the single, calling thread.

File: core/src/main/java/org/neo4j/graphalgo/core/neo4jview/WeightedRelationIteratorImpl.java [hunk 3]
```diff
@@ -1,40 +1,92 @@
...
+        final RelationshipItem relationshipItem = relationships.get();
+        cursor.targetNodeId = graph.toMappedNodeId(relationshipItem.otherNode(originalNodeId));
+        final Cursor<PropertyItem> propertyCursor = relationshipItem.property(propertyKey);
+        if(!propertyCursor.next()) {
+            cursor.weight = defaultWeight;
+            return cursor;
         }
+
+        final PropertyItem propertyItem = propertyCursor.get();
+        cursor.weight = (double) propertyItem.value();
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/neo4jview/WeightedRelationIteratorImpl.java]: use ((Number)propertyItem.value()).doubleValue()

**Comment by knutwalker** [comment]
I think we have a thread safe view in another issue, it's not supposed to be safe yet.

File: core/src/main/java/org/neo4j/graphalgo/core/neo4jview/RelationIteratorImpl.java (hunk 1 above)
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/neo4jview/RelationIteratorImpl.java]: We think it is very uncommon to use one iterator from several threads. I agree with you to document this.

**Pull request merged**